enhancements/removing-midstream.md

```diff
@@ -0,0 +1,322 @@
+to get the latest commit for the `main` branch, we would do:
+
+```bash
+$ git ls-remote https://github.com/tektoncd/pipeline.git main | awk '{ print $1 }'
+07adee706154ce7b79045dd109f1382e997b3163
+```
+
+We can automate this with a `Makefile` target. The idea would be to able to upgrade a component
+in particular, a set of component or all component. Something like the following:
+
+```bash 
+$ make sources/upgrade
+# fetch the latest commit (git ls-remote) for each components
+# mutate upstream_sources.yaml with the result
+$ make sources/pipeline/upgrade
+# fetch the latest commit for pipeline
+# mutate upstream_sources.yaml with the result
+$ make sources/pipeline/upgrade sources/triggers/upgrade
+# fetch the latest commit for pipeline and triggers
+# mutate upstream_sources.yaml with the result
+```
+
+### Read operator metadata and fetch payload versions
+
+For this to work, we need to assume the upstream `tektoncd/operator` repository (or any personal fork)
+hold the payload information somewhere. As of today this is the case in [`test/config.sh`](https://github.com/tektoncd/operator/blob/main/test/config.sh),
+but this is subject to change if we feel like it.
+
+Given that assumption, the goal here would be to fetch the payload from this information and copy the
+content into the correct set of containers.
+
+```bash
+$ make sources/operator/fetch-payload
+# […]
+```
+
+TBD
+
+### Apply patches if need be
+
+All our container builds are done in multi-stage : one the builds and one that is the final image.
+We can use this to our advantage to apply the patches at build time, just after getting the sources in.
+In a nutshell we are looking to do the following (for each container image):
+
+- copy remote source
+- copy patches
+- apply patches
+- build
+
+This means, however that patches needs to live at the same level as the `Dockerfile`. As there is multiple 
+images for a given component, we will have to "duplicate" the `patches` on all component's.
+
+```bash 
+$ make patches
+# sync patches from ./patches to each container images
+$ make container/openshift-pipelines-controller/patches
+# sync patches for the openshift-pipelines-controller image
+```
+
+
+### Build images
+
+Images will be build by `brew` in the Red Hat Toolchain. That said, it should be possible to build images,
+and even the bundle, ahead of time. This is already the case, by doing a similar job as what CPaaS does with
+the templates (`.in` files).
+
+```bash 
+$ make images
+# builds all images but the operator bundle
+$ make images/push
+# builds and push all images but the operator bundle
+$ make images/digest
+# build, push and write image digest for all images but the operator bundle
+$ make container/openshift-pipelines-controller
+# build openshift-pipelines-controller image
+$ make container/openshift-pipelines-controller/push
+# build and push openshift-pipelines-controller image
+# […]
+```
+
```

Comment:
Shall we add a 

### Generate Operator Bundle

and add a make targe to run the bundle generation make target in operator.

Reply:
I could add it in the example, as of today, you can do it the same as the rest or with a "shortcut" (just like `make images`).

```bash
$ make container/openshift-pipelines-operator-bundle
# […]
$ make container/openshift-pipelines-operator-bundle/push
# Or
# […]
$ make bundle
# […]
$ make bundle/push
```